File: modules/loc8-mesh/ios/MeshDeduplicator.swift
//
// MeshDeduplicator.swift
// Loc8Mesh
//
// Thread-safe deduplicator with LRU eviction and time-based expiry.
// Port of bitchat's MessageDeduplicator (The Unlicense — public domain),
// configured LRU 1000 entries / 300 s per spike brief §3.
//
// Dedup key format (must match Android byte-for-byte):
//   "<senderHex(16, lowercase)>-<timestampMs(decimal)>-<type(decimal)>-<first 4 bytes of SHA256(payload), lowercase hex>"
//

import CryptoKit
import Foundation

final class MeshDeduplicator {
    private struct Entry {
        let id: String
        let timestamp: Date
    }

    private var entries: [Entry] = []
    private var head = 0
    private var lookup: [String: Date] = [:]
    private let lock = NSLock()
    private let maxAge: TimeInterval
    private let maxCount: Int

    init(maxAge: TimeInterval = MeshConstants.dedupMaxAgeSeconds,
         maxCount: Int = MeshConstants.dedupMaxCount) {
        self.maxAge = maxAge
        self.maxCount = maxCount
    }

    static func key(for frame: MeshFrame) -> String {
        let senderHex = frame.senderID.map { String(format: "%02x", $0) }.joined()
        let digest = SHA256.hash(data: frame.payload)
        let payloadHash = digest.prefix(4).map { String(format: "%02x", $0) }.joined()
        return "\(senderHex)-\(frame.timestampMs)-\(MeshConstants.messageType)-\(payloadHash)"
    }

    /// Check if the ID was already seen and record it if not.
    func isDuplicate(_ id: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        let now = Date()
        cleanupOldEntries(before: now.addingTimeInterval(-maxAge))

        if lookup[id] != nil { return true }

        entries.append(Entry(id: id, timestamp: now))
        lookup[id] = now
        trimIfNeeded()
        return false
    }

    /// Record an ID without checking (marks our own originations so loopback is dropped).
    func markProcessed(_ id: String) {
        lock.lock()
        defer { lock.unlock() }
        if lookup[id] == nil {
            let now = Date()
            entries.append(Entry(id: id, timestamp: now))
            lookup[id] = now
            trimIfNeeded()
        }
    }

    func reset() {
        lock.lock()
        defer { lock.unlock() }
        entries.removeAll()
        head = 0
        lookup.removeAll()
    }

    private func trimIfNeeded() {
        let activeCount = entries.count - head
        guard activeCount > maxCount else { return }
        // Trim down to 75% of maxCount for amortization (bitchat behavior).
        let targetCount = (maxCount * 3) / 4
        let removeCount = activeCount - targetCount
        for i in head..<(head + removeCount) {
            lookup.removeValue(forKey: entries[i].id)
        }
        head += removeCount
        if head > entries.count / 2 {
            entries.removeFirst(head)
            head = 0
        }
    }

    private func cleanupOldEntries(before cutoff: Date) {
        while head < entries.count, entries[head].timestamp < cutoff {
            lookup.removeValue(forKey: entries[head].id)
            head += 1
        }
        if head > 0 && head > entries.count / 2 {
            entries.removeFirst(head)
            head = 0
        }
    }
}

File: GUARDBACKUP/modules/loc8-mesh/android/src/main/java/expo/modules/loc8mesh/MeshDeduplicator.kt
// MeshDeduplicator.kt
// Loc8Mesh
//
// Thread-safe deduplicator with LRU eviction and time-based expiry — the
// behavioral twin of modules/loc8-mesh/ios/MeshDeduplicator.swift (a port of
// bitchat's MessageDeduplicator, The Unlicense), configured LRU 1000 entries
// / 300 s per spike brief §3.
//
// Dedup key format (must match iOS char-for-char):
//   "<senderHex(16, lowercase)>-<timestampMs(decimal)>-<type(decimal)>-<first 4 bytes of SHA256(payload), lowercase hex>"
// For type 0x30 the decimal segment is "48".

package expo.modules.loc8mesh

import java.security.MessageDigest

class MeshDeduplicator(
    private val maxAgeMs: Long = MeshConstants.DEDUP_MAX_AGE_MS,
    private val maxCount: Int = MeshConstants.DEDUP_MAX_COUNT
) {
    private class Entry(val id: String, val timestampMs: Long)

    private val entries = ArrayDeque<Entry>()
    private val lookup = HashMap<String, Long>()
    private val lock = Any()

    companion object {
        fun key(frame: MeshFrame): String {
            val senderHex = buildString {
                for (b in frame.senderID) append("%02x".format(b.toInt() and 0xFF))
            }
            val digest = MessageDigest.getInstance("SHA-256").digest(frame.payload)
            val payloadHash = buildString {
                for (i in 0 until 4) append("%02x".format(digest[i].toInt() and 0xFF))
            }
            val type = MeshConstants.MESSAGE_TYPE.toInt() and 0xFF // 48
            // toUnsignedString matches Swift's UInt64 decimal formatting even for
            // (already-guarded) > 2^63 wire values.
            val timestamp = java.lang.Long.toUnsignedString(frame.timestampMs)
            return "$senderHex-$timestamp-$type-$payloadHash"
        }
    }

    /** Check if the ID was already seen and record it if not. */
    fun isDuplicate(id: String): Boolean {
        synchronized(lock) {
            val now = System.currentTimeMillis()
            cleanupOldEntries(now - maxAgeMs)

            if (lookup.containsKey(id)) return true

            entries.addLast(Entry(id, now))
            lookup[id] = now
            trimIfNeeded()
            return false
        }
    }

    /** Record an ID without checking (marks our own originations so loopback is dropped). */
    fun markProcessed(id: String) {
        synchronized(lock) {
            if (!lookup.containsKey(id)) {
                val now = System.currentTimeMillis()
                entries.addLast(Entry(id, now))
                lookup[id] = now
                trimIfNeeded()
            }
        }
    }

    fun reset() {
        synchronized(lock) {
            entries.clear()
            lookup.clear()
        }
    }

    private fun trimIfNeeded() {
        if (entries.size <= maxCount) return
        // Trim down to 75% of maxCount for amortization (bitchat behavior).
        val targetCount = (maxCount * 3) / 4
        while (entries.size > targetCount) {
            val removed = entries.removeFirst()
            lookup.remove(removed.id)
        }
    }

    private fun cleanupOldEntries(cutoffMs: Long) {
        while (entries.isNotEmpty() && entries.first().timestampMs < cutoffMs) {
            val removed = entries.removeFirst()
            lookup.remove(removed.id)
        }
    }
}
